fendermint/vm/interpreter/src/fvm/externs.rs

```diff
@@ -29,10 +60,60 @@ impl Consensus for FendermintExterns {
     }
 }
 
-impl Chain for FendermintExterns {
-    fn get_tipset_cid(&self, _epoch: ChainEpoch) -> anyhow::Result<Cid> {
-        unimplemented!("not expecting to use tipsets")
+impl<DB> Chain for FendermintExterns<DB>
+where
+    DB: Blockstore + Clone + 'static,
+{
+    // for retreiving the tipset_cid, we load the chain metadata actor state
+    // at the given state_root and retrieve the blockhash for the given epoch
+    fn get_tipset_cid(&self, epoch: ChainEpoch) -> anyhow::Result<Cid> {
+        // create a read only state tree from the state root
+        let bstore = ReadOnlyBlockstore::new(&self.blockstore);
+        let state_tree = StateTree::new_from_root(&bstore, &self.state_root)?;
```

Comment:
If this was lazily loaded with the `RefCell<Option<...>>` then it would only have to read the root of the state tree once. We can do it as an optimisation later, though.

Reply:
Will create a follow up task